core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java

```diff
@@ -0,0 +1,625 @@
+                if (entry.getValue().size() > 1) {
+                    if (explained.add(entry.getValue())) {
+                        log.warn("Dependencies with duplicate files detected. The dependencies " + entry.getValue()
+                                + " contain duplicate files, e.g. " + entry.getKey());
+                    }
+                }
+            }
+
+            copyCommonContent(runnerZipFs, services, applicationArchivesBuildItem, transformedClasses, generatedClasses,
+                    generatedResources, seen);
+            generateManifest(runnerZipFs, classPath.toString(), packageConfig);
+
+        }
+
+        runnerJar.toFile().setReadable(true, false);
+
+        return new UberJarBuildItem(runnerJar, originalJar);
+    }
+
+    @BuildStep
+    public ThinJarBuildItem buildThinJar(CurateOutcomeBuildItem curateOutcomeBuildItem,
+            OutputTargetBuildItem outputTargetBuildItem,
+            TransformedClassesBuildItem transformedClasses,
+            ApplicationArchivesBuildItem applicationArchivesBuildItem,
+            PackageConfig packageConfig,
+            List<GeneratedClassBuildItem> generatedClasses,
+            List<GeneratedResourceBuildItem> generatedResources) throws Exception {
+        Path thinJarDirectory = outputTargetBuildItem.getOutputDirectory()
+                .resolve(outputTargetBuildItem.getBaseName());
+        IoUtils.recursiveDelete(thinJarDirectory);
+        Files.createDirectories(thinJarDirectory);
+
+        Path runnerJar = thinJarDirectory
+                .resolve(outputTargetBuildItem.getBaseName() + packageConfig.runnerSuffix + ".jar");
+        Path libDir = thinJarDirectory.resolve("lib");
+        Files.createDirectories(libDir);
+
+        try (FileSystem runnerZipFs = ZipUtils.newZip(runnerJar)) {
+
+            log.info("Building thin jar: " + runnerJar);
+
+            final AppModelResolver depResolver = curateOutcomeBuildItem.getResolver();
+            final Map<String, String> seen = new HashMap<>();
+            final StringBuilder classPath = new StringBuilder();
+            final Map<String, List<byte[]>> services = new HashMap<>();
+
+            final List<AppDependency> appDeps = curateOutcomeBuildItem.getEffectiveModel().getUserDependencies();
+
+            copyLibraryJars(transformedClasses, libDir, depResolver, classPath, appDeps);
+            copyCommonContent(runnerZipFs, services, applicationArchivesBuildItem, transformedClasses, generatedClasses,
+                    generatedResources, seen);
+            generateManifest(runnerZipFs, classPath.toString(), packageConfig);
+
+        }
+        runnerJar.toFile().setReadable(true, false);
+        return new ThinJarBuildItem(runnerJar, libDir);
+    }
+
+    /**
+     * Native images are built from a specially created jar file. This allows for changes in how the jar file is generated.
+     *
+     * At present this is the same as the thin jar
```

Comment:
Until they get different, I'd prefer to have a single method that builds the JAR and returns its path that could be used to resolve the lib dir as its sibling.

Reply:
That comment is outdated, they are different. The standard runner jar should not have the native image feature generated.

Reply:
The differences are the file name suffix, the logging and the list of the generated classes, which is passed an argument to the method.
````
            List<GeneratedClassBuildItem> allClasses = new ArrayList<>(generatedClasses);
            allClasses.addAll(substrateResources.stream()
                    .map((s) -> new GeneratedClassBuildItem(true, s.getName(), s.getClassData())).collect(Collectors.toList()));
````
The logic though is identical, isn't it?